### PR TITLE
[AINode] Adding scheduler to support concurrent inference

### PR DESCRIPTION
**Pull request by yunbow30944**
## Description
+ Adding scheduler to support concurrent inference
+ Using basic memory controll logic

### CODE REVIEW

File: iotdb-core/ainode/ainode/core/inference/scheduler/abstract_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+class AbstractScheduler(ABC):
+    """
+    This is the abstract scheduler to schedule inference.
+    """
+
+    def __init__(self, waiting_queue, running_queue, finished_queue):
+        self.waiting_queue = waiting_queue
+        self.running_queue = running_queue
+        self.finished_queue = finished_queue
```
Comment (CRZbulabula) [inline iotdb-core/ainode/ainode/core/inference/scheduler/abstract_scheduler.py]: Append more annotations for this abstract class, including:
1. What is the function of `scheduler`
2. The usage of these three queues
3. The behavior of following interfaces, args... returns....
U can refer to `AbstractInferencePipeline`

File: iotdb-core/ainode/ainode/core/inference/scheduler/basic_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        max_activate_size=10,
+        max_step_size=10,
+    ):
+        super().__init__(waiting_queue, running_queue, finished_queue)
+        self.max_memory_bytes = max_memory_bytes
+        self.max_activate_size = max_activate_size
+        self.max_step_size = max_step_size
+
+    def memory_is_available(self):
+        used = torch.cuda.memory_allocated()  # memory allocated to tensors
```
Comment (CRZbulabula) [inline iotdb-core/ainode/ainode/core/inference/scheduler/basic_scheduler.py]: Which of the follows is the return value of `torch.cuda.memory_allocated()`?
1. The allocated memory of this process
2. The allocated memory of the device where this process resides

File: iotdb-core/ainode/ainode/core/inference/scheduler/basic_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        reserved = (
+            torch.cuda.memory_reserved()
+        )  # memory reserved by the caching allocator
+        # logger.debug(f"Memory used: {used} bytes, Max memory: {self.max_memory_bytes} bytes")
```
Comment (CRZbulabula) [inline iotdb-core/ainode/ainode/core/inference/scheduler/basic_scheduler.py]: U can uncomment this log since it is already in `debug` level. BTW, it's better to follow the log format in `InferenceRequestPool`, which is better for debugging a complex system.

File: iotdb-core/ainode/ainode/core/inference/inference_request_pool.py [hunk 4]
```diff
@@ -65,65 +67,88 @@ def __init__(
...
+            f"Differentiation : {peak_memory_1_token-memory_before_warmup:.2f} bytes"
         )
-        self._running_queue.put(request)
+
+    def _activate_requests(self):
+        requests = self._scheduler.schedule_activate()
+        for request in requests:
+            request.inputs = request.inference_pipeline.preprocess_inputs(
+                request.inputs
+            )
```
Comment (CRZbulabula) [inline iotdb-core/ainode/ainode/core/inference/inference_request_pool.py:113]: U should deliver the scheduled requests to running_queue here rather than within the scheduler. Otherwise, some requests would be `step` without preprocess their inputs.

File: iotdb-core/ainode/ainode/core/inference/inference_request_pool.py [hunk 5]
```diff
@@ -150,3 +177,10 @@ def run(self):
 
     def stop(self):
         self._stop_event.set()
+        logger.info(f"[Pool-{self.pool_id}] Stopping and releasing GPU memory.")
```
Comment (CRZbulabula) [inline iotdb-core/ainode/ainode/core/inference/inference_request_pool.py]: Add `Device` tag for this log, the same as below

File: iotdb-core/ainode/ainode/core/inference/inference_request_pool.py [hunk 6]
```diff
@@ -65,65 +67,88 @@ def __init__(
...
-            self.device
-        )  # Ensure output tensor is on the same device
-        request.write_step_output(output[0].mean(dim=0))
-        request.inference_pipeline.post_decode()
-        if request.is_finished():
-            request.inference_pipeline.post_inference()
-            logger.debug(
-                f"[Inference][Device-{self.device}][Pool-{self.pool_id}][ID-{request.req_id}] Request is finished"
+        for request in requests:
+            inputs = request.inputs.to(self.device)
```
Comment (CRZbulabula) [inline iotdb-core/ainode/ainode/core/inference/inference_request_pool.py]: Be cautious to this line, seems it is redundant when the `step` is executed for multiple times.

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
PTAL

File: iotdb-core/ainode/ainode/core/inference/scheduler/basic_scheduler.py (hunk 2 above)
Comment (yunbow30944) [reply iotdb-core/ainode/ainode/core/inference/scheduler/basic_scheduler.py]: The allocated memory of this process , i.e. the inference request pool , but it only includes memory allocated to tensors.
The total GPU memory reserved by the PyTorch caching allocator for this process (including memory held for reuse), can be obtained via `torch.cuda.memory_reserved()` .

File: iotdb-core/ainode/ainode/core/inference/scheduler/basic_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import torch
+
+from ainode.core.inference.inference_request import InferenceRequest
+from ainode.core.inference.scheduler.abstract_scheduler import AbstractScheduler
+from ainode.core.log import Logger
+
+logger = Logger()
+
+
+class BasicScheduler(AbstractScheduler):
```
Comment (CRZbulabula) [inline iotdb-core/ainode/ainode/core/inference/scheduler/basic_scheduler.py:31]: Append a log to briefly introduce the strategy of this scheduler

**Review by CRZbulabula — APPROVED** [review]
LGTM!
